lib/backend.ml

```diff
@@ -300,3 +367,84 @@ module Regression = struct
     let f = Field.to_filename in
     Fields.to_list ~query_responses:f ~commands:f ~logs:f
 end
+
+module Benchmark = struct
+  type t = {
+    mutable bench : Benchmark_summary.t;
+    mutable query_responses : Query_response.t list;
+    mutable commands : Command.t list;
+    mutable logs : Logs.t list;
+    merlins : Merlin.t list;
+  }
+  [@@deriving fields]
+
+  let kind = Bench
+
+  let create_initial merlins =
+    {
+      bench = { name = "Merlin benchmark"; results = StringMap.empty };
+      query_responses = [];
+      commands = [];
+      logs = [];
+      merlins;
+    }
+
+  let persist_logs ~log tables = tables.logs <- log :: tables.logs
+
+  let all_files () =
+    let f = Field.to_filename in
+    Fields.to_list ~bench:f ~query_responses:f ~commands:f ~logs:f ~merlins:f
+
+  let dump ~dump_dir t =
+    let d = dump_dir in
+    let () =
+      Fields.iter
+        ~bench:(Field.dump_single Benchmark_summary.pp d t)
+        ~query_responses:(Field.dump Query_response.pp d t)
+        ~commands:(Field.dump Command.pp d t)
+        ~logs:(Field.dump Logs.pp d t) ~merlins:(Field.dump Merlin.pp d t)
+    in
+    ()
+
+  let update_analysis_data ~id ~responses ~cmd ~file
+      ~loc:(_loc : Import.location) ~merlin_id ~query_type tables =
+    let max_timing, _timings, responses =
+      (* FIXME: add json struture to the two lists *)
+      let rec loop ~max_timing ~responses ~timings = function
+        | [] -> (max_timing, timings, responses)
+        | resp :: rest ->
+            let timing = Merlin.Response.get_timing resp in
+            let timings = timing :: timings in
+            let responses = resp :: responses in
+            let max_timing = Int.max timing max_timing in
+            loop ~max_timing ~timings ~responses rest
+      in
+      loop ~max_timing:Int.min_int ~responses:[] ~timings:[] responses
+    in
+    let resp =
+      (* TODO: make a cli-argument out of this instead of doing this always *)
+      let responses = List.map Merlin.Response.crop_value responses in
+      { Query_response.sample_id = id; responses; merlin_id }
+    in
+    let cmd = { Command.sample_id = id; cmd; merlin_id } in
+    let filename = Fpath.filename (Fpath.v (File.filename file)) in
+    let metric =
+      {
+        Benchmark_metric.name = Merlin.Query_type.to_string query_type;
+        value = [ max_timing ];
+        units = "ms";
+        description = "";
+        trend = None;
+      }
+    in
+    let upd = function
+      | Some x -> Some (Benchmark_result.update x metric)
+      | None -> Some (Benchmark_result.create filename metric)
+    in
+    let result = StringMap.update filename upd tables.bench.results in
```

Comment:
As mentioned: I'd accumulate all metrics among different files instead of having one `Benchmark_metric.t` per filename.